src/ast/rewriter/bit_blaster/bit_blaster_adder.cpp

```diff
@@ -0,0 +1,112 @@
+#include "ast/rewriter/bit_blaster/bit_blaster_adder.h"
+
+bit_blaster_adder::bit_blaster_adder(bool_rewriter & rewriter, unsigned sz, numeral const & value):
+    m_rewriter(rewriter),
+    m_constant(value)
+{
+    reduce();
+    expr_ref_vector empty(m());
+    m_variable.resize(sz, empty);
+}
+
+bit_blaster_adder::bit_blaster_adder(bool_rewriter & rewriter, unsigned sz, expr * const * bits):
+    bit_blaster_adder(rewriter, sz)
+{
+    for (unsigned i = 0; i < sz; i++)
+        add_bit(i, bits[i]);
+}
+
+expr_ref bit_blaster_adder::sum_bits(vector< expr_ref_vector > & columns, expr_ref_vector & out_bits) const {
+    SASSERT(out_bits.empty());
+
+    expr_ref_vector carries(m());
+    expr_ref tmp1(m()), tmp2(m()), tmp3(m());
+
+    for (auto & column : columns) {
+        column.append(carries);
+        carries.reset();
+
+        while (column.size() >= 3) {
+            expr * a = column.back();
+            column.pop_back();
+            expr * b = column.back();
+            column.pop_back();
+            expr * c = column.back();
+            column.pop_back();
+
+            m_rewriter.mk_xor(a, b, tmp1);
+            m_rewriter.mk_xor(tmp1, c, tmp2);
+            column.push_back(tmp2);
+
+            m_rewriter.mk_and(a, b, tmp2);
+            m_rewriter.mk_and(tmp1, c, tmp3);
+            // tmp2 and tmp3 can't be true at the same time, so use
+            // whichever of mk_or vs mk_xor makes the most sense here.
+            m_rewriter.mk_or(tmp2, tmp3, tmp1);
+            carries.push_back(tmp1);
+        }
+
+        if (column.size() == 2) {
+            expr * a = column.back();
+            column.pop_back();
+            expr * b = column.back();
+            column.pop_back();
+
+            m_rewriter.mk_xor(a, b, tmp1);
+            column.push_back(tmp1);
+
+            m_rewriter.mk_and(a, b, tmp1);
+            carries.push_back(tmp1);
+        }
+
+        out_bits.push_back(column.get(0, m().mk_false()));
```

Comment:
correct, but subtle.
Why not explicit case on sz?

switch (sz) {
case 0:
case 1:
case 2:
default:
....
}